xlators/cluster/afr/src/afr-open.c

```diff
@@ -271,34 +295,223 @@ afr_fd_ctx_need_open(fd_t *fd, xlator_t *this, unsigned char *need_open)
+               "Failed getlk for %s", uuid_utoa(local->fd->inode->gfid));
+    }
+
+    if (xdata)
+        local->replies[i].xdata = dict_ref(xdata);
+
+    call_count = afr_frame_return(frame);
+
+    if (call_count)
+        return 0;
+
+    /* Currently we get 3 values from the lower layer (protocol/client) in the
+     * getlk_cbk.
+     *  FD_REOPEN_ALLOWED : No conflicting locks are held and reopen is allowed
+     *  FD_REOPEN_NOT_ALLOWED : Conflicting locks are held and reopen is not
+     *                          allowed
+     *  FD_BAD : FD is not valid
+     *
+     * - If we get FD_REOPEN_NOT_ALLOWED from any of the bricks, will block the
+     *   reopen taking this as high priority.
+     * - If we get FD_BAD from all the replies, we will not reopen since we do
+     *   not know the correct status.
+     * - If we get FD_BAD from few brick and FD_REOPEN_NOT_ALLOWED from one or
+     *   more bricks, then we will block reopen.
+     * - If we get FD_BAD from few bricks and FD_REOPEN_ALLOWED from one or
+     *   more bricks, then we will allow the reopen.
+     *
+     *   We will update the final_reopen_status only when the value returned
+     *   from lower layer is >= FD_REOPEN_ALLOWED and < FD_BAD. We will not set
+     *   FD_BAD in final_reopen_status, since it can lead to unexpected
+     *   behaviours.
+     *
+     *   At the end of this loop, if we still have final_reopen_status as -1
+     *   i.e., the init value, it means we failed to get the fd status from any
+     *   of the bricks or we do not have a valid fd on any of the bricks. We
+     *   will not reopen the fd in this case as well.
+     */
+
+    for (i = 0; i < priv->child_count; i++) {
+        if (final_reopen_status != FD_REOPEN_NOT_ALLOWED &&
+            local->replies[i].xdata) {
+            ret = dict_get_int32(xdata, "fd-reopen-status", &fd_reopen_status);
+            if (ret) {
+                gf_msg(this->name, GF_LOG_ERROR, -ret, AFR_MSG_DICT_GET_FAILED,
+                       "Failed to get whether reopen is allowed or not on fd "
+                       "for file %s on subvolume %s.",
+                       local->loc.path, priv->children[i]->name);
+            } else if (fd_reopen_status >= FD_REOPEN_ALLOWED &&
+                       fd_reopen_status < FD_BAD) {
+                final_reopen_status = fd_reopen_status;
+            }
+        }
+
+        if (final_reopen_status == FD_REOPEN_NOT_ALLOWED)
+            break;
+    }
+
+    if (final_reopen_status == FD_REOPEN_NOT_ALLOWED) {
+        gf_log(this->name, GF_LOG_INFO,
+               "Conflicting locks held on file %s. FD reopen is not allowed.",
+               local->loc.path);
+    } else if (final_reopen_status == -1) {
+        gf_log(this->name, GF_LOG_INFO,
+               "Failed to get the lock information "
+               "on file %s. FD reopen is not allowed.",
+               local->loc.path);
+    } else {
+        afr_do_fix_open(frame, this);
```

Comment:
Before calling this, it is better to afr_local_replies_wipe() to avoid potential dict leaks. We currently don't overwrite it in afr_openfd_fix_open_cbk(), but it is good to future proof it.

Reply:
Done.